s3driver.go

```diff
@@ -0,0 +1,216 @@
+			size:  *o.Size,
+			mtime: *o.LastModified,
+		})
+	}
+	for _, o := range objects.CommonPrefixes {
+		files = append(files, &fileInfo{
+			name: strings.TrimSuffix(strings.TrimPrefix(*o.Prefix, prefix), "/"),
+			mode: os.ModeDir,
+		})
+	}
+	return files, nil
+}
+
+func (d S3Driver) DeleteDir(path string) error {
+	translatedPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return err
+	}
+	_, err = d.s3.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(d.bucket),
+		Key:    aws.String(translatedPath),
+	})
+	return err
+}
+
+func (d S3Driver) DeleteFile(path string) error {
+	translatedPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return err
+	}
+	_, err = d.s3.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(d.bucket),
+		Key:    aws.String(translatedPath),
+	})
+	return err
+}
+
+func (d S3Driver) Rename(oldpath string, newpath string) error {
+	translatedOldpath, err := translatePath(d.homePath, oldpath)
+	if err != nil {
+		return err
+	}
+	translatedNewpath, err := translatePath(d.homePath, newpath)
+	if err != nil {
+		return err
+	}
+
+	if _, err := d.s3.CopyObject(&s3.CopyObjectInput{
+		Bucket:     aws.String(d.bucket),
+		CopySource: aws.String(d.bucket + "/" + translatedOldpath),
+		Key:        &translatedNewpath,
+	}); err != nil {
+		return err
+	}
+
+	if _, err = d.s3.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(d.bucket),
+		Key:    &translatedOldpath,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d S3Driver) MakeDir(path string) error {
+	localPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return err
+	}
+	if !strings.HasSuffix(localPath, "/") {
+		localPath += "/"
+	}
+
+	_, err = d.s3.PutObject(&s3.PutObjectInput{
+		Bucket: aws.String(d.bucket),
+		Key:    aws.String(localPath),
+		Body:   bytes.NewReader([]byte{}),
+	})
+	return err
+}
+
+func (d S3Driver) GetFile(path string) (io.ReadCloser, error) {
+	localPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return nil, err
+	}
+	obj, err := d.s3.GetObject(&s3.GetObjectInput{
+		Bucket: aws.String(d.bucket),
+		Key:    aws.String(localPath),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return obj.Body, nil
+}
+
+func (d S3Driver) PutFile(path string, r io.Reader) error {
+	localPath, err := translatePath(d.homePath, path)
+	if err != nil {
+		return err
+	}
+
+	rawData, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	_, err = d.s3.PutObject(&s3.PutObjectInput{
+		Bucket: aws.String(d.bucket),
+		Key:    aws.String(localPath),
+		Body:   bytes.NewReader(rawData),
+	})
+	return err
+}
+
+func translatePath(prefix, path string) (s3Path string, err error) {
+	cleanPath := filepath.Clean("/" + prefix + filepath.Clean("/"+path))
+	// For some reason, filepath.Clean drops trailing /'s, so if there was one we have to put it back
+	if strings.HasSuffix(path, "/") {
+		cleanPath += "/"
+	}
+	return strings.TrimLeft(cleanPath, "/"), nil
+}
```

Comment:
That's understandable; I had to re-write it about 6 times. I'll try to come up with a comment to explain what it's doing ;p